### DISCUSSION TITLE
PP Algorithm

### DISCUSSION THREAD

**Discussion by IVII1** [ppy/osu#28692] [Ideas]
Hello osu! devs (and fellow players!)

I recently heard that a new PP update is on its way which will remove combo scaling. I think this is an amazing change because it will make things that are actually important, like accuracy and number of mistakes (misses + sliderbreaks as Willy called them) contribute way more to the final PP value of each play

A concern i heard from other players is while theoretically this will make identical plays (in terms of accuracy and mistakes) be worth exactly the same, it will not take into account weather the player missed on an easy or a hard part within the map.

I'm not sure if this problem has been discussed among people in charge of PP, but i would like to offer a solution to this problem.

Make it so each map is split into smaller "map chunks" which act as maps within a map, granting PP on their own based on the same metrics mentioned above. Now that combo scaling is removed. this should work properly since before it's removal this would be a bad solution due to each "map chunk" having low max combo, and i see no issues other than that with this method. Obviously let me know if there is something i missed regarding this.

Another reason this will work well is because some of these chunks will obviously contain hard parts of a map, and some will contain the easy parts. 

Without map chunks, every note is treated equally so a one miss during the easy part is the same during the hard part. The map chunk system would ensure each mistake is interpreted according to the map chunk it's in (which would take into account the speed and the spacing and whatever else it already does, those parts of the PP algorithm would remain unchanged)

The only problem i see with this is accuracy, since if you split map into small maps with a small amount of object, each imperfectly hit object would have a big impact of the accuracy within that small map chunk, possibly tanking its PP value

If this ends up being a problem, you could make it so the accuracy part of the PP value is added as a modifier at the end of the map.

Let me know your thoughts, questions and concerns if there's any, thank you for reading and have a nice day/night.

EDIT: spelling

**Comment by cihe13375** [comment]
I think the main problem is that most scores don't have their replays saved on the server (even if they do, processing all these replays would likely require significant computing resources), so it's impossible to tell the number of misses in each chunk, or generally "if the player missed in easy parts".

**Reply by IVII1** [reply]
I guess you would have to process it on the players PC then, and while somehow mitigating potential exploits that might come due to the calculation being done on the client PC (possibly cheat engine interference)

Though this might be stoppable since you can just ban players whose pp values change without any reworks or setting new plays

You could also set it up so every replay is sent to the server, parsed and then deleted. but that would probably be a LOT of writing and deleting files, so i don't know if that's plausible

I did completely forget that not all replays are saved on the server which makes it impossible to fully implement this

You could save combo before a break as a way to keep track of chunks. So if, for example, a map has 1000 combo and the player has 3 misses you could save the combo that player held before each miss, and compare that amount of combo at that point in the map to the actual notes in the map to distinguish between hard and easy parts

**Reply by Tenexxt** [reply]
After the pp update, most scores will most likely be incompatible anyway, as they are probably changing not only combo scailing.

**Comment by Tenexxt** [comment]
I think a good idea for this is to just make this into kiai. Idk what kiai does, but I'm guessing that in kiai you get more points (might be wrong, never cared enough to check), so making a chorus, aka, The most Likely harder part give more pp, since you are already getting more points wouldn't even be a good idea, it just makes sense for it to do that.

**Reply by IVII1** [reply]
yeah but the issue with that is as of right now the algorithm is only getting the ending screen and the map itself to try to deduce how much pp you should get. It has no concept of kiais, or difficulty of parts within a map, only the overall map difficulty (as far as im aware, correct me if I'm wrong)

That's why i think a good solution would be to save the combo before each miss/slider break and send that info to the algorithm as well

Every time combo resets, you can be sure the player made a mistake and you can deduce weather they missed or sliderbroke by the combo leading up on that mistake

So if, for example, a player has 300 and the 301st combo is a mistake, then you can check weather that 301st combo (the combo where the mistake happened) was a circle or a slider head (in this case it would be a miss) or a slider end/tail (this would be a slider break) 

The player keeps playing and the algorithm tracks the current combo until other misses, and the same logic can apply

When PP is being evaluated, the algorithm would look at how hard achieving that combo before the miss was and award pp accordingly.

You can then add all the pp gained from all the combos together to get a final pp value for the whole play

The only issue i see with this is that plays set before this change would not be treated exactly the same as the plays after due to the new information being tracked now (unless they already tracked this information before)
